### PR TITLE
Add --no-index to pip_conf fixture to reduce testing time

### LINKED ISSUE #919: Add --no-index to pip_conf fixture to reduce testing time
#### What's the problem this feature will solve?

Currently pytest runs offline tests surprisingly slow if the network is down. See:

```bash
$ time pytest -m "not network'
...

real    2m5.040s
user    0m3.728s
sys    0m0.719s
```

I suspect that is because pip_conf fixture has --index-url=http://example.com and --trusted-host=example.com options and it looks like pip spends much time on hitting the fake PyPI server.

Adding --no-index option to the pip_conf makes tests faster:

```bash
$ time pytest -m 'not network' -k 'not (test_default_pip_conf_read or test_command_line_overrides_pip_conf or test_extra_index_option or test_default_index_url)'
...

real    0m3.736s
user    0m3.118s
sys    0m0.560s
```

_(4 tests are skipped intentionally, for they assert https://example.com)._

📈 The result is ~2 min vs ~4 sec.

#### Describe the solution you'd like

1. Update pip_conf fixture to use no-index option:

```ini
[global]
no-index = true
find-links = {wheels_path}
```

2. Create new fixture foo_index_pip_conf (Any suggestions to the name? Bikeshedding please.):

```ini
[global]
index-url = http://example.com
```

Note --trusted-host is removed.

3. Update the following tests to use new fixture:
    * test_default_pip_conf_read
    * test_command_line_overrides_pip_conf
    * test_extra_index_option
    * test_default_index_url

#### Alternative Solutions

Removing --trusted-host option makes also offline tests fast, but it looks hacky and I'd prefer to set up --no-index explicitly on tests which not require PyPI server.

#### Additional context

N/A

### PR DESCRIPTION
**Pull request by hramezani**
Closes #919.

**Changelog-friendly one-liner**: Add `--no-index` to pip_conf fixture to reduce testing time.

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

File: tests/test_cli_compile.py [hunk 1]
```diff
@@ -23,11 +23,34 @@
...
+    monkeypatch.setenv("PIP_CONFIG_FILE", path)
+
+    try:
+        yield path
+    finally:
+        os.remove(path)
+
+
+@pytest.fixture
+def pip_with_index_conf(tmpdir, monkeypatch):
```
Comment (atugushev) [inline tests/test_cli_compile.py]: I'd prefer to create a helper fixture to make code more DRY:
```python
@pytest.fixture
def make_pip_conf(tmpdir, monkeypatch):
    created_paths = []

    def _make_pip_conf(content):
        pip_conf_file = "pip.conf" if os.name != "nt" else "pip.ini"
        path = (tmpdir / pip_conf_file).strpath

        with open(path, "w") as f:
            f.write(content)

        monkeypatch.setenv("PIP_CONFIG_FILE", path)

        created_paths.append(path)
        return path

    try:
        yield _make_pip_conf
    finally:
        for path in created_paths:
            os.remove(path)
```
and then create needed fixtures by:

```python
@pytest.fixture
def pip_conf(make_pip_conf):
    return make_pip_conf(
        dedent(
            """\
            [global]
            no-index = true
            find-links = {wheels_path}
            """.format(
                wheels_path=MINIMAL_WHEELS_PATH
            )
        )
    )

@pytest.fixture
def pip_with_index_conf(make_pip_conf):
    return make_pip_conf(
        dedent(
            """\
            [global]
            index-url = http://example.com
            find-links = {wheels_path}
            """.format(
                wheels_path=MINIMAL_WHEELS_PATH
            )
        )
    )
```

Comment (hramezani) [reply tests/test_cli_compile.py]: @atugushev Good advice. It is Done!

**Comment by hramezani** [comment]
@atugushev Should I press the merge button? It's green now :smile:

**Comment by atugushev** [comment]
I've tried this PR locally and got failed `test_pre_option`.

**Comment by atugushev** [comment]
@hramezani try to turn off the network and `run pytest -k 'not network'`?

**Comment by atugushev** [comment]
Looks like `small-fake-a==0.3b1` depends on `six==1.10.0`. We could fix it in the following pull request.

**Comment by atugushev** [comment]
@hramezani 

> Should I press the merge button? It's green now 😄

Yes, you may merge this now.

**Comment by hramezani** [comment]
@atugushev I merged this and will create another PR for https://github.com/jazzband/pip-tools/pull/934#issuecomment-541096618

**Comment by atugushev** [comment]
@hramezani thanks!

**Pull request merged**